src/java.base/share/classes/java/io/SerializationMisdeclarationChecker.java

```diff
@@ -0,0 +1,259 @@
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package java.io;
+
+import java.lang.reflect.Field;
+import java.lang.reflect.Member;
+import java.lang.reflect.Method;
+import java.lang.reflect.Proxy;
+import java.security.AccessController;
+import java.security.PrivilegedAction;
+import java.util.Arrays;
+
+import static jdk.internal.event.SerializationMisdeclarationEvent.*;
+import static java.lang.reflect.Modifier.*;
+
+final class SerializationMisdeclarationChecker {
+
+    private static final String SUID_NAME = "serialVersionUID";
+    private static final String SERIAL_PERSISTENT_FIELDS_NAME = "serialPersistentFields";
+    private static final String WRITE_OBJECT_NAME = "writeObject";
+    private static final String READ_OBJECT_NAME = "readObject";
+    private static final String READ_OBJECT_NO_DATA_NAME = "readObjectNoData";
+    private static final String WRITE_REPLACE_NAME = "writeReplace";
+    private static final String READ_RESOLVE_NAME = "readResolve";
+
+    private static final Class<?>[] WRITE_OBJECT_PARAM_TYPES = {ObjectOutputStream.class};
+    private static final Class<?>[] READ_OBJECT_PARAM_TYPES = {ObjectInputStream.class};
+    private static final Class<?>[] READ_OBJECT_NO_DATA_PARAM_TYPES = {};
+    private static final Class<?>[] WRITE_REPLACE_PARAM_TYPES = {};
+    private static final Class<?>[] READ_RESOLVE_PARAM_TYPES = {};
+
+    @SuppressWarnings("removal")
+    static void checkMisdeclarations(Class<?> cl) {
+        AccessController.doPrivileged((PrivilegedAction<Object>) () -> {
+            privilegedCheckSerialVersionUID(cl);
+            privilegedCheckSerialPersistentFields(cl);
+
+            privilegedCheckPrivateMethod(cl, WRITE_OBJECT_NAME,
+                    WRITE_OBJECT_PARAM_TYPES, Void.TYPE);
+            privilegedCheckPrivateMethod(cl, READ_OBJECT_NAME,
+                    READ_OBJECT_PARAM_TYPES, Void.TYPE);
+            privilegedCheckPrivateMethod(cl, READ_OBJECT_NO_DATA_NAME,
+                    READ_OBJECT_NO_DATA_PARAM_TYPES, Void.TYPE);
+
+            privilegedCheckAccessibleMethod(cl, WRITE_REPLACE_NAME,
+                    WRITE_REPLACE_PARAM_TYPES, Object.class);
+            privilegedCheckAccessibleMethod(cl, READ_RESOLVE_NAME,
+                    READ_RESOLVE_PARAM_TYPES, Object.class);
+
+            return null;
+        });
+    }
+
+    private static void privilegedCheckSerialVersionUID(Class<?> cl) {
+        Field f = declaredField(cl, SUID_NAME);
+        if (f == null) {
+            if (isOrdinaryClass(cl)) {
+                commitEvent(cl, SUID_NAME + " should be declared explicitly" +
+                        " as a private static final long field");
+            }
+            return;
+        }
+        if (cl.isEnum()) {
+            commitEvent(cl, SUID_NAME + " in an enum class is not effective");
```

Comment:
Yes, that's perhaps clearer, will do.

Reply:
Typically in other places in the JDK we use `priviledgedXxx` for naming methods that are simple wrappers that call `xxx` from within a `doPrivileged`. The method is called privileged because it doesn't require the caller to use `doPrivileged`.

That is something like:

```
String privilegedGetProperty(String property) {
     return AccessController.doPrivileded((...) () -> System.getProperty(property));
} 
```

See for instance: https://github.com/openjdk/jdk/blob/ee98d262181f5822609674c71c85ad4576ac1632/src/java.base/share/classes/sun/security/action/GetPropertyAction.java#L107

Reply:
Ah OK, I wasn't aware of this convention.